### ISSUE TITLE
[v2.1.1] Crash at opal_convertor_generic_simple_position while sending/receiving GPU memory with multi-threads at comm_dup channel

### ISSUE THREAD

**Issue by woso** [open-mpi/ompi#4610]
### background

We are building a machine learning system with support of async training basing on OMPI

In my simple test case, With OMPI, worker send batch  of keys to server（pull）, then server probe the message length and the rank, and do something else final send values back(process), at the same time worker is ready to receive it; 

To support a node is both worker and server, we use MPI_comm_dup to separate the channel into upstream and downstream; And with the purpose of training, GPU is use as the default mode of calculation and communication; 

The simple case will crash after hundred of iterations; 

![image](https://user-images.githubusercontent.com/504521/33871985-d7be0c54-df4f-11e7-91d0-69b2b5964940.png)

![image](https://user-images.githubusercontent.com/504521/33872002-f660ba4e-df4f-11e7-8c23-fc7a13e142fb.png)

### OMPI install

download openmpi-2.1.1.tar.gz
./configure -with-hwloc=internal --with-cuda=/usr/local/cuda-8.0 --enable-mpi-thread-multiple --prefix=/home/a/openmpi/

### the system 

* Operating system/version: 
CentOS 7.2  
* Computer hardware: 
Intel Xeon E5-2682 v4 * 2 (hyper to 64 threads) memory 256G;  Nvidia P100 *2
* Network type: 
ethernet tcp/ip with 2 card: eth4 and eth5 and binding to bond0 for communication

-----------------------------

###  build & run the case

```shell
shell$ mpirun -np 2 --mca plm_rsh_no_tree_spawn 1 --mca btl self,sm,tcp --mca btl_tcp_if_include bond0 -hostfile XXX ./all2all_pure_test --gtest_filter='All2AllTest.GPU*' 500000 1
```

**Comment by woso** [comment]
#include <mpi.h>
#include <cuda.h>
#include <cuda_runtime.h>
#include <thread>

#include "gtest/gtest.h"

enum TestType {
  TCPU,        // use cpu at both sides
  TGPU,        // use gpu at both sides
  TPIN,        // use pin at both sides
  TCOPY,       // copy gpu to cpu before send/recv
};

static unsigned iterations = 1;
static unsigned type = 0;
static float scale = 1.0f;

namespace xdl {
namespace ams {

#define TAG_EXIT (1<<20)

#ifdef NDEBUG
    #undef NDEBUG
#endif
#include <assert.h>

#define CHECK(x) assert((x))
#define CUDA_CHECK(x) \
  do { \
    cudaError_t error = (x); \
    CHECK(error == cudaSuccess); \
  } while (0)

class All2AllTest: public ::testing::Test {
 public:
  static void SetUpTestCase() {
    C *= scale;

    std::cout << "iterations: " << iterations << std::endl;
    std::cout << "sparse param count: " << C << std::endl;

    int provided, ret;
    ret = MPI_Init_thread(nullptr, nullptr, MPI_THREAD_MULTIPLE, &provided);
    CHECK(ret == MPI_SUCCESS);

    ret = MPI_Comm_size(MPI_COMM_WORLD, &world_size_);
    CHECK(ret == MPI_SUCCESS);

    ret = MPI_Comm_rank(MPI_COMM_WORLD, &rank_);
    CHECK(ret == MPI_SUCCESS);

    int rank = -1;
    ret = MPI_Comm_dup(MPI_COMM_WORLD, &upstream_);
    CHECK(ret == MPI_SUCCESS);
    ret = MPI_Comm_dup(MPI_COMM_WORLD, &downstream_);
    CHECK(ret == MPI_SUCCESS);

    ret = MPI_Comm_rank(upstream_, &rank);
    CHECK(ret == MPI_SUCCESS && rank == rank_);
    ret = MPI_Comm_rank(downstream_, &rank);
    CHECK(ret == MPI_SUCCESS && rank == rank_);

    ASSERT_TRUE(world_size_ > 0);
    ASSERT_TRUE(rank_ >= 0 && rank_ < world_size_);

    std::cout << "[" << rank_ << " / " << world_size_ << "] " << std::endl;

    send_reqs_.resize(world_size_);
    recv_reqs_.resize(world_size_);

  }

  static void TearDownTestCase() {
    CHECK(MPI_Finalize() == MPI_SUCCESS);
  }

  /* called by TEST */
  static bool Iterate(unsigned n, TestType t);

  static size_t D;
  static size_t C;

  static int world_size_;
  static int rank_;
  static std::vector<MPI_Request> send_reqs_;
  static std::vector<MPI_Request> recv_reqs_;
 private:
  static unsigned iter_;
  static unsigned exit_;

  static MPI_Comm upstream_;    // down -> upper
  static MPI_Comm downstream_;  // upper -> down

  /* for agent send */
  static uint64_t *keys_a_;
  static float *val_a_;

  /* for server recv */
  static uint64_t *keys_s_;
  static float *val_s_;

  /* for agent */
  static bool Pull(const uint64_t *key);
  static bool PullAck(float *val);

  /* for server */
  static bool Ack(const float *val, size_t len, int rank, int tag);
  static bool Process(uint64_t *key, const float *val);
  static bool Loop(uint64_t *keys, float *values);
};

size_t All2AllTest::D = 18;
size_t All2AllTest::C = 300000;

int All2AllTest::world_size_;
int All2AllTest::rank_;
MPI_Comm All2AllTest::upstream_;
MPI_Comm All2AllTest::downstream_;

std::vector<MPI_Request> All2AllTest::send_reqs_;
std::vector<MPI_Request> All2AllTest::recv_reqs_;

uint64_t *All2AllTest::keys_a_;
float *All2AllTest::val_a_;

uint64_t *All2AllTest::keys_s_;
float *All2AllTest::val_s_;

unsigned All2AllTest::iter_ = 0;
unsigned All2AllTest::exit_ = 0;

/// XXX: make-sure rbuf is avalid until PullAck call Wait
bool All2AllTest::Pull(const uint64_t *key) {
  cudaSetDevice(0);
  unsigned kcount = C / world_size_;
  for (size_t k = 0; k < world_size_; ++k) {
    //MPI_Isend((const char *)(key + kcount * k), sizeof(uint64_t) * kcount, MPI_BYTE, k, iter_, upstream_, &send_reqs_[k]);
    MPI_Isend((const char *)(key), sizeof(uint64_t) * kcount, MPI_BYTE, k, iter_, upstream_, &send_reqs_[k]);
  }
  //std::cout << rank_ << " pull " << iter_ << std::endl;
  return true;
}

bool All2AllTest::PullAck(float *val) {
  cudaSetDevice(0);
  unsigned vcount = D*C / world_size_;
  for (size_t k = 0; k < world_size_; ++k) {
    //MPI_Irecv((char *)(val + vcount * k), sizeof(float) * vcount, MPI_BYTE, k, iter_, downstream_, &recv_reqs_[k]);
    MPI_Irecv((char *)(val), sizeof(float) * vcount, MPI_BYTE, k, iter_, downstream_, &recv_reqs_[k]);
  }

  for (size_t k = 0; k < world_size_; ++k) {
    MPI_Status status;
    CHECK(MPI_Wait(&send_reqs_[k], &status) == MPI_SUCCESS);
    CHECK(MPI_Wait(&recv_reqs_[k], &status) == MPI_SUCCESS);
  }
  //std::cout << rank_ << " pula " << iter_ << std::endl;
  return true;
}

bool All2AllTest::Ack(const float *val, size_t len, int rank, int tag) {
  cudaSetDevice(0);
  MPI_Status status;
  MPI_Send((char *)(val), len, MPI_BYTE, rank, tag, downstream_);
  //std::cout << "[" << rank << "<-" << rank_ << "] on pula " << tag << std::endl;
  return true;
}

/// return false on exit
bool All2AllTest::Process(uint64_t *key, const float *val) {
  cudaSetDevice(0);
  MPI_Status status;
  int ret = MPI_Probe(MPI_ANY_SOURCE, MPI_ANY_TAG, upstream_, &status);
  CHECK(ret == MPI_SUCCESS);
  int len, tag, rank;
  rank = status.MPI_SOURCE;
  tag = status.MPI_TAG;
  MPI_Get_count(&status, MPI_BYTE, &len);

  if (tag == TAG_EXIT) {
    std::cout << "[" << rank << "->" << rank_ << "] on exit" << std::endl;
    MPI_Recv((char *)(val), len, MPI_BYTE, rank, tag, upstream_, &status);
    if ((++exit_) == world_size_) {
      return false;
    }
    return true;
  }

  unsigned kcount = len / sizeof(uint64_t);
  unsigned vcount = D * kcount;
  CHECK(kcount <= C);
  CHECK(vcount <= D*C);

  MPI_Recv((char *)(key), len, MPI_BYTE, rank, tag, upstream_, &status);

  if (tag % 100 == 0) {
    std::cout << "[" << rank << "->" << rank_ << "] on pull " << tag << std::endl;
  }

  //Ack(val, len, rank, tag);
  auto t = std::thread(Ack, val, sizeof(float) * vcount, rank, tag);
  t.detach();

  return true;
}

bool All2AllTest::Loop(uint64_t *keys, float *val) {
  cudaSetDevice(0);
  std::cout << rank_ << " startup" << std::endl;
  while (Process(keys, val)) { }
  std::cout << rank_ << " shutdown" << std::endl;
  return true;
}

bool All2AllTest::Iterate(unsigned n, TestType t) {
  cudaSetDevice(0);
  iter_ = 0;
  exit_ = 0;

  switch (t) {
  case TGPU:
  case TCOPY:
       CUDA_CHECK(cudaMalloc(&keys_s_, sizeof(uint64_t)*C));
       CUDA_CHECK(cudaMalloc(&val_s_, sizeof(float)*D*C));
       CUDA_CHECK(cudaMalloc(&keys_a_, sizeof(uint64_t)*C));
       CUDA_CHECK(cudaMalloc(&val_a_, sizeof(float)*D*C));
       break;
  case TCPU:
       keys_s_ = (uint64_t *)malloc(sizeof(uint64_t)*C);
       val_s_ = (float *)malloc(sizeof(float)*D*C);
       keys_a_ = (uint64_t *)malloc(sizeof(uint64_t)*C);
       val_a_ = (float *)malloc(sizeof(float)*D*C);
       break;
  case TPIN:
       assert(!"no implement");
       break;
  }
CHECK(nullptr != keys_s_);
  CHECK(nullptr != val_s_);
  CHECK(nullptr != keys_a_);
  CHECK(nullptr != val_a_);

  std::thread serv = std::thread(Loop, keys_s_, val_s_);

  for (size_t it = 0; it < n && iter_ < TAG_EXIT; ++it) {
    CHECK(Pull(keys_a_));
    CHECK(PullAck(val_a_));

    iter_ ++;
  }

  for (size_t k = 0; k < world_size_; ++k) {
    MPI_Send(nullptr, 0, MPI_BYTE, k, TAG_EXIT, upstream_);
    std::cout << rank_ << " exit" << std::endl;
  }

  serv.join();
  return true;
}

TEST_F(All2AllTest, GPUTest) {
  cudaSetDevice(0);
  Iterate(iterations, TGPU);
}

TEST_F(All2AllTest, CPUTest) {
  cudaSetDevice(0);
  Iterate(iterations, TCPU);
}

TEST_F(All2AllTest, PINTest) {
  cudaSetDevice(0);
  Iterate(iterations, TCPU);
}

TEST_F(All2AllTest, COPYTest) {
  cudaSetDevice(0);
  Iterate(iterations, TCOPY);
}

}  // namespace ams
}  // namespace xdl

int main(int argc, char **argv)
{
  cudaSetDevice(0);
  ::testing::InitGoogleTest(&argc, argv);

  for (int i = 1; i < argc; ++i) {
    printf("arg %2d = %s\n", i, argv[i]);
  }

  if (argc > 1) {
    iterations = atoi(argv[1]);
  }

  if (argc > 2) {
    scale = atof(argv[2]);
  }

  return RUN_ALL_TESTS();
}

**Comment by jsquyres** [comment]
@sjeaugey Can you (or someone at NVIDIA) have a look at this?

**Comment by woso** [comment]
The test case would crash immediately while using GPUTest; 
IF the key count 'C' is less or not call std::thread for Ack but directly call Ack in same thread, 
the test case will work well.

**Comment by sjeaugey** [comment]
The crash seems to happen in the TCP BTL (mca_btl_tcp_endpoint_recv_handler). It would be good to compile MPI with debug symbols to get a better trace and precise information about why we get a crash.

It's not clear to me either if that has anything to do with GPU communication -- @woso would you have an easy way to not use GPUs and check if it still crashes ? I don't see anything CUDA specific in the back trace ; but it could be that we crash trying to access CUDA memory.

**Comment by woso** [comment]
> would you have an easy way to not use GPUs and check if it still crashes ?
    
By using CPU memory for communication, as the case 'CPUTest', allocating mem by malloc instead of cudaMalloc, the test case will work well.

> It would be good to compile MPI with debug symbols to get a better trace and precise information about why we get a crash
 
OK, I'll try

**Comment by sjeaugey** [comment]
Yes, 1.10.7 is a relatively old (but stable) version of Open MPI.

Also, just to check CUDA support is there and working, could you launch your test with "-mca btl self,smcuda,tcp" ?

**Comment by zycao-naoc** [comment]
I'm working together with @woso . I think the crash came out from the fragmented receiving when CUDA support is enabled, after doing some experiments and tracing around the GPU case.

Notice the backtrace info at the top of this thread, the code reported 'address not mapped' when getting info 'opal_convertor_set_position_nocheck()' function. In my experiments, this function would not be called when everything was OK. Once it was called, the crash came out immediately.

Furthermore, calling 'opal_convertor_set_position' was done in the macro 'MCA_PML_OB1_RECV_REQUEST_UNPACK', which is used 'mca_pml_ob1_recv_request_progress_*' (including 'mca_pml_ob1_recv_request_progress_frag'). I added an output for the 'data_offset' in this macro. It happened to find that when error occurs, two different data receiving streams from 2 threads of same process became overlapping. As the picture showed below:
![image](https://user-images.githubusercontent.com/10669111/33974157-e94e19a0-e0c1-11e7-87d2-a43f34a1573c.png)

Normally, it should be like this:
![image](https://user-images.githubusercontent.com/10669111/33974235-4c6fce3e-e0c2-11e7-8989-efff58d973ce.png)

I'm not able to get deeper tracing for the fragment queuing code. But I could imagine that if all data packages were not fragmented, the crash could be avoid. Thus I tried to set btl_smcuda_cuda_max_send_size and btl_self_max_send_size with very big value (24000000 for this case). Then the test could run smoothly. However, I guess this setting would do harm to the communication performance.

I hope this information would help to check out the reason of how fragmenting receiving would lead to the crash,  and wish to see this issue could be eventually fixed.

**Comment by sjeaugey** [comment]
Thanks for the information. So I understand you're running with the smcuda BTL (instead of sm) and is still crashes. Is that right ?

Also, did you try 1.10.7 ?

**Comment by Stonesjtu** [comment]
I had a similar error when using `smcuda` with multi-thread. There is also a server with 2 threads to receive data from two workers, and it works fine with CPU memory but fails with GPU memory. So I'm wondering if the `smcuda` component is thread-safe.

traceback:
```

[date:05402] Signal: Segmentation fault (11)
[date:05402] Signal code: Address not mapped (1)
[date:05402] Failing at address: 0x70
[date:05402] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x10330)[0x7f8504ed7330]
[date:05402] [ 1] /usr/local/lib/openmpi/mca_btl_smcuda.so(mca_btl_smcuda_component_progress+0x3c1)[0x7f84ba2ee811]
[date:05402] [ 2] /usr/local/lib/libopen-pal.so.13(opal_progress+0x2a)[0x7f84c8c1bcfa]
[date:05402] [ 3] /usr/local/lib/openmpi/mca_pml_ob1.so(mca_pml_ob1_recv+0x54d)[0x7f84b9ab982d]
[date:05402] [ 4] /usr/local/lib/libmpi.so.12(MPI_Recv+0x12f)[0x7f84e8eeed0f]
```


Envs:
- ubuntu 16.04LTS
- open-mpi: 1.10.7 (build with `../configure --with-cuda --enable-mpi-thread-multiple`)
- 1080Ti X 2
